appinventor/appengine/src/com/google/appinventor/client/actions/DeleteAction.java

```diff
@@ -25,33 +25,32 @@ public void execute() {
       @Override
       public void execute() {
         if (Ode.getInstance().getCurrentView() == Ode.PROJECTS) {
-          List<Project> selectedProjects =
-              ProjectListBox.getProjectListBox().getProjectList().getSelectedProjects();
+          List<Project> selectedProjects = ProjectListBox.getProjectListBox().getProjectList()
+              .getSelectedProjects();
           List<ProjectFolder> selectedFolders = ProjectListBox.getProjectListBox().getProjectList().getSelectedFolders();
-          if (selectedProjects.size() > 0 || selectedFolders.size() > 0) {
-            List<Project> projectsToDelete = selectedProjects;
+          if (!selectedProjects.isEmpty() || !selectedFolders.isEmpty()) {
+            List<Project> projectsToDelete = new ArrayList<>(selectedProjects);
+            List<ProjectFolder> foldersToDelete = new ArrayList<>(selectedFolders);
             for (ProjectFolder f : selectedFolders) {
               projectsToDelete.addAll(f.getNestedProjects());
+              foldersToDelete.addAll(f.getNestedFolders());
             }
             // Show one confirmation window for selected projects.
-            if (deleteConfirmation(projectsToDelete)) {
-              for (Project project : projectsToDelete) {
-                project.moveToTrash();
-              }
-              for (ProjectFolder f : selectedFolders) {
-                f.getParentFolder().removeChildFolder(f);
-              }
+            if (deleteConfirmation(true, projectsToDelete, foldersToDelete)) {
+              Ode.getInstance().getFolderManager().moveItemsToFolder(selectedProjects, selectedFolders,
+                  Ode.getInstance().getFolderManager().getTrashFolder());
+              Ode.getInstance().getFolderManager().saveAllFolders();
             }
           } else {
             // The user can select a project to resolve the
             // error.
             ErrorReporter.reportInfo(MESSAGES.noProjectSelectedForDelete());
           }
         } else { //We are deleting a project in the designer view
-          List<Project> selectedProjects = new ArrayList<Project>();
+          List<Project> selectedProjects = new ArrayList<>();
           Project currentProject = Ode.getInstance().getProjectManager().getProject(Ode.getInstance().getCurrentYoungAndroidProjectId());
           selectedProjects.add(currentProject);
-          if (deleteConfirmation(selectedProjects)) {
+          if (deleteConfirmation(true, selectedProjects, new ArrayList<>())) {
```

Comment:
It might be better to use `Collections.emptyList()` if GWT supports it in order to not have to result a new list every time this gets called.